### PR TITLE
Add a 'replay' feature

### LINKED ISSUE #22545: Add TestHarness feature to re-play previous results
## Reason
Long runs with a short terminal buffer are problematic. For this reason, allowing a re-play feature which will also make use of other options (such as --failed-tests, --re) will be a boon for these situations.

## Design
Leverage the already useful .previous_results feature to display back results as they appeared in std out previously missed.

## Impact
Allows running the TestHarness again to play back a previous run (without actually running anything)

### PR DESCRIPTION
**Pull request by milljm**
Allow users to reprint the previous results:

```bash
./run_tests --show-last-run
```

Compatible with additional test spec selection options (--re, -i, --failed-tests).

Closes #22545

### CODE REVIEW

**Comment by GiudGiud** [comment]
love the idea

**Comment by GiudGiud** [comment]
so I got this output running it before any tests were run
```
(moose) giudgl@fn654105:~/projects/moose_v4/test$ ./run_tests --show-last-run
runWorker Exception: Traceback (most recent call last):
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/schedulers/Scheduler.py", line 412, in runJob
    self.run(job) # Hand execution over to derived scheduler
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/schedulers/RunParallel.py", line 38, in run
    job_results = self.options.results_storage[job.getTestDir()][job.getTestName()]
KeyError: '/Users/giudgl/projects/moose_v4/test/tests/misc/check_error'

Traceback (most recent call last):
  File "/Users/giudgl/projects/moose_v4/test/./run_tests", line 8, in <module>
    TestHarness.buildAndRun(sys.argv, None, MOOSE_DIR)
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/TestHarness.py", line 181, in buildAndRun
    harness.findAndRunTests()
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/TestHarness.py", line 434, in findAndRunTests
    self.scheduler.waitFinish()
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/schedulers/Scheduler.py", line 178, in waitFinish
    self.__sortAndLaunch()
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/schedulers/Scheduler.py", line 172, in __sortAndLaunch
    self.queueJobs(jobs, j_lock)
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/schedulers/Scheduler.py", line 255, in queueJobs
    self.status_pool.apply_async(self.jobStatus, (job, jobs, j_lock))
  File "/Users/giudgl/mambaforge3/envs/moose/lib/python3.10/multiprocessing/pool.py", line 458, in apply_async
    self._check_running()
  File "/Users/giudgl/mambaforge3/envs/moose/lib/python3.10/multiprocessing/pool.py", line 353, in _check_running
    raise ValueError("Pool not running")
ValueError: Pool not running
```

and after running the test suite then cancelling it midway

```
(moose) giudgl@fn654105:~/projects/moose_v4/test$ ./run_tests --re meshgen -j8
meshgenerators/sideset_around_subdomain_generator.group/adjacent_to_boundary ... [MESH_MODE!=DISTRIBUTED] SKIP
meshgenerators/xy_delaunay_generator.xydelaunay/basic ..................................................... OK
meshgenerators/sidesets_bounding_box_generator.generate/generate_sidesets_bounding_box_test ............... OK
meshgenerators/combiner_generator.translations/single ..................................................... OK
meshgenerators/break_mesh_by_block_generator.break_mesh_by_blocks_generator/3d_auto_test .................. OK
meshgenerators/advanced_extruder_generator.test ........................................................... OK
meshgenerators/block_deletion_generator.block_deletion/all_by_block_ids ................................... OK
meshgenerators/rename_block_generator.rename/all_ids ...................................................... OK
meshgenerators/generated_mesh_generator.test .............................................................. OK
meshgenerators/annular_mesh_generator.annular_mesh_generator .............................................. OK
meshgenerators/fill_between_point_vectors_generator.bow ................................................... OK
meshgenerators/file_mesh_generator.test ................................................................... OK
meshgenerators/rename_boundary_generator.rename/all_ids ................................................... OK
meshgenerators/distributed_rectilinear/generator.mesh_generator ........................................... OK
meshgenerators/subdomain_bounding_box_generator.subdomain_bounding_box_generator_test ..................... OK
meshgenerators/mesh_extruder_generator.extruder/basic ..................................................... OK
meshgenerators/fill_between_sidesets_generator.squares .................................................... OK
meshgenerators/sidesets_between_subdomains_generator.group/adjacent_subs .................................. OK
meshgenerators/sideset_around_subdomain_generator.group/inside ............................................ OK
meshgenerators/move_node_generator.generate/single_element ................................................ OK
meshgenerators/final_generator.final_select/first ......................................................... OK
meshgenerators/patch_mesh_generator.group/quad4 ........................................................... OK
meshgenerators/meta_data_store.meta_data_store ............................................................ OK
meshgenerators/polyline_mesh_generator.polyline/basic ..................................................... OK
meshgenerators/errors.tests/get_mesh_invalid .............................................................. OK
meshgenerators/parsed_generate_sideset.parsed_generate_sideset_test ....................................... OK
meshgenerators/extra_nodeset_generator.generate_extra_nodeset ............................................. OK
meshgenerators/refine_block_generator.generate/single_block ............................................... OK
meshgenerators/transform_generator.rotate_and_scale ....................................................... OK
meshgenerators/bounding_box_nodeset_generator.gen_nodesets/inside ......................................... OK
meshgenerators/refine_sideset_generator.generate/left ..................................................... OK
meshgenerators/cartesian_mesh_generator.cartesian_mesh/1D ................................................. OK
meshgenerators/break_boundary_on_subdomain.boundary_type/external ......................................... OK
meshgenerators/stack_generator.stack/2d ................................................................... OK
meshgenerators/plane_id_mesh_generator.2d_cartesian_grid/x_dir ............................................ OK
meshgenerators/explode_mesh_generator.all/2D .............................................................. OK
meshgenerators/image_mesh_generator.image_mesh_generator_test ............................................. OK
meshgenerators/element_subdomain_id_generator.assign/quad_test ............................................ OK
meshgenerators/lower_d_block_generator.mesh/ids ........................................................... OK
meshgenerators/element_generator.generate/single_element .................................................. OK
meshgenerators/concentric_circle_mesh_generator.mesh_only ................................................. OK
meshgenerators/distributed_rectilinear/dmg_displaced_mesh.adaptivity ...................................... OK
meshgenerators/compute_node_position.2d ................................................................... OK
meshgenerators/boundary_deletion_generator.boundary_deletion_test ......................................... OK
meshgenerators/add_all_side_sets_generators.simple ........................................................ OK
meshgenerators/stitched_mesh_generator.stitched_mesh_generator/mesh_only .................................. OK
meshgenerators/parsed_subdomain_mesh_generator.parsed_subdomain_mg_test ................................... OK
meshgenerators/cyclic.error ............................................................................... OK
meshgenerators/sidesets_by_normals_generator.generate_sidesets_normals_test ............................... OK
meshgenerators/ringleb_mesh_generator.ringleb_mesh_generator_test ......................................... OK
meshgenerators/unique_extra_id_mesh_generator.test/default ................................................ OK
^C
Exiting due to keyboard interrupt...
meshgenerators/sidesets_from_normals_generator.generate/normals_follow .................................... OK
runWorker Exception: Traceback (most recent call last):
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/schedulers/Scheduler.py", line 438, in runJob
    self.queueJobs(jobs, j_lock)
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/schedulers/Scheduler.py", line 255, in queueJobs
    self.status_pool.apply_async(self.jobStatus, (job, jobs, j_lock))
  File "/Users/giudgl/mambaforge3/envs/moose/lib/python3.10/multiprocessing/pool.py", line 458, in apply_async
    self._check_running()
  File "/Users/giudgl/mambaforge3/envs/moose/lib/python3.10/multiprocessing/pool.py", line 353, in _check_running
    raise ValueError("Pool not running")
ValueError: Pool not running

runWorker Exception: Traceback (most recent call last):
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/schedulers/Scheduler.py", line 438, in runJob
    self.queueJobs(jobs, j_lock)
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/schedulers/Scheduler.py", line 255, in queueJobs
    self.status_pool.apply_async(self.jobStatus, (job, jobs, j_lock))
  File "/Users/giudgl/mambaforge3/envs/moose/lib/python3.10/multiprocessing/pool.py", line 458, in apply_async
    self._check_running()
  File "/Users/giudgl/mambaforge3/envs/moose/lib/python3.10/multiprocessing/pool.py", line 353, in _check_running
    raise ValueError("Pool not running")
ValueError: Pool not running
^[[AFatal Python error: _enter_buffered_busy: could not acquire lock for <_io.BufferedWriter name='<stdout>'> at interpreter shutdown, possibly due to daemon threads
Python runtime state: finalizing (tstate=0x00000001536043f0)

Current thread 0x0000000104674580 (most recent call first):
  <no Python frame>

Extension modules: yaml._yaml, numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg._umath_linalg, numpy.fft._pocketfft_internal, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator, pandas._libs.tslibs.np_datetime, pandas._libs.tslibs.dtypes, pandas._libs.tslibs.base, pandas._libs.tslibs.nattype, pandas._libs.tslibs.timezones, pandas._libs.tslibs.tzconversion, pandas._libs.tslibs.ccalendar, pandas._libs.tslibs.fields, pandas._libs.tslibs.timedeltas, pandas._libs.tslibs.timestamps, pandas._libs.properties, pandas._libs.tslibs.offsets, pandas._libs.tslibs.parsing, pandas._libs.tslibs.conversion, pandas._libs.tslibs.period, pandas._libs.tslibs.vectorized, pandas._libs.ops_dispatch, pandas._libs.missing, pandas._libs.hashtable, pandas._libs.algos, pandas._libs.interval, pandas._libs.tslib, pandas._libs.lib, pandas._libs.hashing, pandas._libs.ops, pandas._libs.arrays, pandas._libs.index, pandas._libs.join, pandas._libs.sparse, pandas._libs.reduction, pandas._libs.indexing, pandas._libs.internals, pandas._libs.writers, pandas._libs.window.aggregations, pandas._libs.window.indexers, pandas._libs.reshape, pandas._libs.tslibs.strptime, pandas._libs.groupby, pandas._libs.testing, pandas._libs.parsers, pandas._libs.json, hit (total: 56)
Abort trap: 6
(moose) giudgl@fn654105:~/projects/moose_v4/test$ ./run_tests --show-last-run
runWorker Exception: Traceback (most recent call last):
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/schedulers/Scheduler.py", line 412, in runJob
    self.run(job) # Hand execution over to derived scheduler
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/schedulers/RunParallel.py", line 38, in run
    job_results = self.options.results_storage[job.getTestDir()][job.getTestName()]
KeyError: '/Users/giudgl/projects/moose_v4/test/tests/misc/check_error'

Traceback (most recent call last):
  File "/Users/giudgl/projects/moose_v4/test/./run_tests", line 8, in <module>
    TestHarness.buildAndRun(sys.argv, None, MOOSE_DIR)
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/TestHarness.py", line 181, in buildAndRun
    harness.findAndRunTests()
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/TestHarness.py", line 434, in findAndRunTests
    self.scheduler.waitFinish()
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/schedulers/Scheduler.py", line 178, in waitFinish
    self.__sortAndLaunch()
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/schedulers/Scheduler.py", line 172, in __sortAndLaunch
    self.queueJobs(jobs, j_lock)
  File "/Users/giudgl/projects/moose_v4/python/TestHarness/schedulers/Scheduler.py", line 255, in queueJobs
    self.status_pool.apply_async(self.jobStatus, (job, jobs, j_lock))
  File "/Users/giudgl/mambaforge3/envs/moose/lib/python3.10/multiprocessing/pool.py", line 458, in apply_async
    self._check_running()
  File "/Users/giudgl/mambaforge3/envs/moose/lib/python3.10/multiprocessing/pool.py", line 353, in _check_running
    raise ValueError("Pool not running")
ValueError: Pool not running
```

maybe it should error a bit more gracefully if we send bad data in

**Comment by milljm** [comment]
I can write up something to prevent it from running if there is no results file to deal with. But as for the Ctlr-C in the middle, that has always happened.

File: python/TestHarness/TestHarness.py [hunk 1]
```diff
@@ -972,6 +972,11 @@ def useExistingStorage(self):
         if (os.path.exists(self.options.results_file)
             and (self.options.failed_tests or self.options.pbs or self.options.show_last_run)):
             return True
+        elif ((self.options.failed_tests or self.options.show_last_run)
```
Comment (milljm) [inline python/TestHarness/TestHarness.py:975]: I wasn't sure what to do about PBS... I'd rather not mess with that logic. I _believe_ we are generating one for the user if it does not exist and they are launching PBS jobs for the first time...

**Review by lindsayad — APPROVED** [review]

**Pull request merged**